zio-interop-reactivestreams/src/main/scala/zio/interop/reactivestreams/Adapters.scala

```diff
@@ -3,304 +3,457 @@ package zio.interop.reactivestreams
 import org.reactivestreams.Publisher
 import org.reactivestreams.Subscriber
 import org.reactivestreams.Subscription
+import org.reactivestreams.Processor
 import zio._
 import zio.Unsafe._
 import zio.internal.RingBuffer
 import zio.stream._
-import zio.stream.ZStream.Pull
 
-import java.util.concurrent.atomic.AtomicBoolean
 import java.util.concurrent.atomic.AtomicReference
+import zio.stream.internal.AsyncInputConsumer
+import zio.stream.internal.AsyncInputProducer
+import java.util.concurrent.atomic.AtomicBoolean
+import zio.UIO
+import scala.util.control.NoStackTrace
 
 object Adapters {
 
   def streamToPublisher[R, E <: Throwable, O](
     stream: => ZStream[R, E, O]
-  )(implicit trace: Trace): ZIO[R, Nothing, Publisher[O]] =
-    ZIO.runtime.map { runtime => subscriber =>
-      if (subscriber == null) {
-        throw new NullPointerException("Subscriber must not be null.")
-      } else
-        unsafe { implicit unsafe =>
-          val subscription = new DemandTrackingSubscription(subscriber)
-          runtime.unsafe.fork(
-            for {
-              _ <- ZIO.succeed(subscriber.onSubscribe(subscription))
-              _ <- stream
-                     .run(demandUnfoldSink(subscriber, subscription))
-                     .catchAll(e => ZIO.succeed(subscriber.onError(e)))
-                     .forkDaemon
-            } yield ()
-          )
-          ()
-        }
-    }
+  )(implicit trace: Trace): ZIO[R, Nothing, Publisher[O]] = ZIO.runtime[R].map { runtime => subscriber =>
```

Comment:
This guy cannot be implemented on top of channelToPublisher due to the missing R <: Scope constraint